Source/NSFontCollection.m

```diff
@@ -0,0 +1,341 @@
+{
+  self = [super init];
+  if (self != nil)
+    {
+      _fonts = [[NSMutableArray alloc] initWithCapacity: 50];
+      _queryDescriptors = [[NSMutableArray alloc] initWithCapacity: 10];
+      _exclusionDescriptors = [[NSMutableArray alloc] initWithCapacity: 10];
+      _queryAttributes = [[NSMutableArray alloc] initWithCapacity: 10];
+    }
+  return self;
+}
+
+- (void) dealloc
+{
+  RELEASE(_fonts);
+  RELEASE(_queryDescriptors);
+  RELEASE(_exclusionDescriptors);
+  RELEASE(_queryAttributes);
+  [super dealloc];
+}
+
+// This method will get the actual list of fonts
+- (void) _runQueryWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSEnumerator *en = [queryDescriptors objectEnumerator];
+  GSFontEnumerator *fen = [GSFontEnumerator sharedEnumerator];
+  id d = nil;
+
+  ASSIGNCOPY(_queryDescriptors, queryDescriptors);
+  while ((d = [en nextObject]) != nil)
+    {
+      NSArray *names = [fen availableFontNamesMatchingFontDescriptor: d];
+      id name = nil;
+
+      en = [names objectEnumerator];
+      while ((name = [en nextObject]) != nil)
+        {
+          NSFont *font = [NSFont fontWithName: name size: 0.0]; // get default size
+          [_fonts addObject: font];
+        }
+    }
+}
+
++ (NSFontCollection *) fontCollectionWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSFontCollection *fc = [[NSFontCollection alloc] init];
+  ASSIGNCOPY(fc->_queryDescriptors, queryDescriptors);
+  return fc;
+}
+
++ (NSFontCollection *) fontCollectionWithAllAvailableDescriptors
+{
+  return [self fontCollectionWithDescriptors:
+                 [[GSFontEnumerator sharedEnumerator] availableFontDescriptors]];
+}
+
++ (NSFontCollection *) fontCollectionWithLocale: (NSLocale *)locale
+{
+  return [self fontCollectionWithAllAvailableDescriptors];
+}
+
++ (BOOL) showFontCollection: (NSFontCollection *)collection
+                   withName: (NSFontCollectionName)name
+                 visibility: (NSFontCollectionVisibility)visibility
+                      error: (NSError **)error
+{
+  NSNumber *v = [NSNumber numberWithInt: visibility];
+  [__sharedFontCollections setObject: collection
+                             forKey: name];
+  [__sharedFontCollectionsVisibility setObject: v
+                                        forKey: name];
+  return YES;
+}
+
++ (BOOL) hideFontCollectionWithName: (NSFontCollectionName)name
+                         visibility: (NSFontCollectionVisibility)visibility
+                              error: (NSError **)error
+{
+  
+  [__sharedFontCollectionsHidden addObject: name];
+  return YES;
+}
+
++ (BOOL) renameFontCollectionWithName: (NSFontCollectionName)aname
+                           visibility: (NSFontCollectionVisibility)visibility
+                               toName: (NSFontCollectionName)name
+                                error: (NSError **)error
+{
+  NSFontCollection *fc = [__sharedFontCollections objectForKey: aname];
+  [__sharedFontCollections setObject: fc forKey: name];
+  return YES;
```

Comment:
This method does not set the visibility of the font collection. Nor does it remove the old name from any of the collections.